base/ca/src/main/java/com/netscape/cmscore/dbs/CertificateRepository.java

```diff
@@ -604,49 +602,35 @@ public BigInteger getLastSerialNumberInRange(BigInteger serial_low_bound, BigInt
             logger.debug("CertificateRepository:getLastCertRecordSerialNo: returning " + ret);
             return ret;
         }
-        int ltSize = recList.getSizeBeforeJumpTo();
-
-        logger.debug("CertificateRepository:getLastSerialNumberInRange: ltSize " + ltSize);
-
-        CertRecord curRec = null;
-
-        int i;
-        Object obj = null;
-
-        for (i = 0; i < 5; i++) {
-            obj = recList.getCertRecord(i);
 
-            if (obj != null) {
-                curRec = (CertRecord) obj;
+        CertRecord curRec = iRecs.next();
 
-                BigInteger serial = curRec.getSerialNumber();
+        while (iRecs.hasNext()) {
+            CertRecord followRec = iRecs.next();
 
-                logger.debug("CertificateRepository:getLastCertRecordSerialNo:  serialno  " + serial);
+            BigInteger curSerial = curRec.getSerialNumber();
+            BigInteger followSerial = followRec.getSerialNumber();
 
-                if (((serial.compareTo(serial_low_bound) == 0) || (serial.compareTo(serial_low_bound) == 1)) &&
-                        ((serial.compareTo(serial_upper_bound) == 0) || (serial.compareTo(serial_upper_bound) == -1))) {
-                    logger.debug("getLastSerialNumberInRange returning: " + serial);
-                    if (modeChange && mEnableRandomSerialNumbers) {
-                        mCounter = serial.subtract(serial_low_bound).add(BigInteger.ONE);
-                        logger.debug("getLastSerialNumberInRange mCounter: " + mCounter);
-                    }
-                    return serial;
+            if (followSerial.compareTo(serial_upper_bound) > 0 &&
+                    curSerial.compareTo(serial_upper_bound) < 0) {
```

Comment:
IIUC the upper bound should be considered within the range, so this probably should  be:
```
if (followSerial.compareTo(serial_upper_bound) > 0 &&
        curSerial.compareTo(serial_upper_bound) <= 0) {
```